### PR TITLE
Feature/Add document links handling to labelstudio format

### PR DESCRIPTION
**Pull request by serereg**
- [x] support of document_links
	- [x] import links (it is not clear how to convert links: use file names instead of file_id? If so it involves already working UI)
	- [x] export links (it is needed to modify export endpoint, because document links exist in manifest.json

https://github.com/epam/badgerdoc/issues/66#issuecomment-1405027788

### CODE REVIEW

**Review by spiridonovfed — CHANGES_REQUESTED** [review]

File: convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py [hunk 1]
```diff
@@ -80,6 +86,9 @@ def convert_relation_from_bd(
                 result_items.append(item)
         return result_items
 
+    def convert_document_links_from_bd(self, manifest: Manifest):
```
Comment (cakeinsauce) [inline convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py:89]: return typing, please

File: convert/src/label_studio_to_badgerdoc/models/bd_manifest_model_practic.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+class Manifest(BaseModel):
+    revision: str
+    job_id: int
+    user: Optional[str]
+    pipeline: Optional[Any]
+    date: Optional[str]
+    pages: Dict[str, str]
+    failed_validation_pages: List = []
+    validated: List = []
```
Comment (cakeinsauce) [inline convert/src/label_studio_to_badgerdoc/models/bd_manifest_model_practic.py:22]: Please, specify list of what type it is? E.g. `List[int]`

File: convert/src/label_studio_to_badgerdoc/models/bd_manifest_model_practic.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    links_json: List[DocumentLink]
+    file: str
+    bucket: str
+    categories: List = []
```
Comment (cakeinsauce) [inline convert/src/label_studio_to_badgerdoc/models/bd_manifest_model_practic.py:26]: Please, specify list of what type it is? E.g. `List[int]`

File: convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py [hunk 4]
```diff
@@ -30,10 +34,12 @@ def from_badgerdoc(
         relations = self.convert_relation_from_bd(
             badgerdoc_annotations, badgerdoc_tokens.objs
         )
+        document_links = self.convert_document_links_from_bd(badgerdoc_manifest) if badgerdoc_manifest else []
         item = ModelItem(
             annotations=[Annotation(result=objs + relations)],
             predictions=[],
             data=Data(text=text),
+            meta=Meta(labels=[], relations=document_links)
```
[the remark this answers is not in the document]
Comment (serereg) [reply convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py:42]: This logic is about export, and it is the first time we provide it, so right now here is nothing bad.
Exporting of labels is in different issue, we'll add logic later.

File: convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py [hunk 5]
```diff
@@ -80,6 +86,9 @@ def convert_relation_from_bd(
                 result_items.append(item)
         return result_items
 
+    def convert_document_links_from_bd(self, manifest: Manifest):
+        return [DocumentRelation(**document_link.dict()) for document_link in manifest.links_json]
+
     def form_link_direction(self, link: AnnotationLink) -> str:
         # TODO: add logic
```
[the remark this answers is not in the document]
Comment (serereg) [reply convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py]: Done

File: convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py (hunk 1 above)
Comment (serereg) [reply convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py]: Done

Comment (serereg) [reply convert/src/label_studio_to_badgerdoc/labelstudio_format/label_studio_format.py]: Done

File: convert/src/label_studio_to_badgerdoc/models/bd_manifest_model_practic.py (hunk 3 above)
Comment (serereg) [reply convert/src/label_studio_to_badgerdoc/models/bd_manifest_model_practic.py]: Done

File: convert/src/label_studio_to_badgerdoc/models/label_studio_models.py [hunk 6]
```diff
@@ -101,14 +102,25 @@ class Data(BaseModel):
...
+
 class ModelItem(BaseModel):
     id: int = 1
     annotations: List[Annotation]
     file_upload: str = ""
     drafts: List = []
     predictions: List[int] = []
     data: Data
-    meta: Dict[str, Any] = {}
+    meta: Meta = Meta(labels=[], relations=[])
```
[the remark this answers is not in the document]
Comment (serereg) [reply convert/src/label_studio_to_badgerdoc/models/label_studio_models.py]: Done

File: convert/src/label_studio_to_badgerdoc/badgerdoc_to_label_studio_use_case.py [hunk 7]
```diff
@@ -28,45 +30,49 @@ def execute(
...
+            input_manifest = self.download_file_from_s3(s3_input_manifest, tmp_dir)
 
             page = Page.parse_file(input_tokens)
             annotation = AnnotationConverterToTheory(practic_annotations=bd_annotation_model_practic.BadgerdocAnnotation.parse_file(input_annotations)).convert()
-            return BadgerdocData(page=page, annotation=annotation)
+            manifest = Manifest.parse_file(input_manifest)
+            return BadgerdocData(page=page, annotation=annotation, manifest=manifest)
+
+    def download_file_from_s3(self, s3_path: S3Path, tmp_dir: Path):
+        local_file = tmp_dir / Path(s3_path.path).name
```
[the remark this answers is not in the document]
Comment (serereg) [reply convert/src/label_studio_to_badgerdoc/badgerdoc_to_label_studio_use_case.py]: Done

**Comment by serereg** [comment]
Duplicates https://github.com/epam/badgerdoc/pull/322

**Pull request closed without merging**